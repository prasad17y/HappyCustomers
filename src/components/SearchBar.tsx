import React from 'react';
import {View, TextInput, StyleSheet, Image} from 'react-native';

interface SearchBarProps {
  value: string;
  onChangeText: (text: string) => void;
  placeholder?: string;
}

const SearchBar: React.FC<SearchBarProps> = ({
  value,
  onChangeText,
  placeholder = 'Search',
}) => {
  return (
    <View style={styles.container}>
      <Image
        source={{
          uri: 'https://img.icons8.com/ios-glyphs/30/a0a0a0/search--v1.png',
        }}
        style={styles.icon}
      />
      <TextInput
        style={styles.input}
        value={value}
        onChangeText={onChangeText}
        placeholder={placeholder}
        placeholderTextColor="#8e8e93"
        clearButtonMode="while-editing"
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: '#f0f2f5',
    borderRadius: 10,
    marginHorizontal: 16,
    marginBottom: 8,
    paddingHorizontal: 10,
    height: 40,
  },
  icon: {
    width: 20,
    height: 20,
    marginRight: 8,
  },
  input: {
    flex: 1,
    fontSize: 16,
    color: '#000',
  },
});

export default React.memo(SearchBar);
