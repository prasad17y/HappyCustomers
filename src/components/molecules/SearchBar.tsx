import React, {useState, useEffect, useRef} from 'react';
import {View, TextInput, StyleSheet, StyleProp, ViewStyle} from 'react-native';
import Icon from '../atoms/Icon';

interface SearchBarProps {
  value: string;
  onChangeText: (text: string) => void;
  placeholder?: string;
  debounceDelay?: number;
  style?: StyleProp<ViewStyle>;
}

const SearchBar: React.FC<SearchBarProps> = ({
  value,
  onChangeText,
  placeholder = 'Search',
  debounceDelay = 0,
  style,
}) => {
  const [internalText, setInternalText] = useState(value);
  const timeoutRef = useRef<ReturnType<typeof setTimeout> | null>(null);

  useEffect(() => {
    setInternalText(value);
  }, [value]);

  const handleChangeText = (text: string) => {
    setInternalText(text);

    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
    }

    if (debounceDelay > 0) {
      timeoutRef.current = setTimeout(() => {
        onChangeText(text);
      }, debounceDelay);
    } else {
      onChangeText(text);
    }
  };

  useEffect(() => {
    return () => {
      if (timeoutRef.current) {
        clearTimeout(timeoutRef.current);
      }
    };
  }, []);

  return (
    <View style={[styles.container, style]}>
      <Icon source={require('../../assets/search.png')} style={styles.icon} />
      <TextInput
        style={styles.input}
        value={internalText}
        onChangeText={handleChangeText}
        placeholder={placeholder}
        placeholderTextColor="#8e8e93"
        clearButtonMode="while-editing"
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: '#f0f2f5',
    borderRadius: 10,
    paddingHorizontal: 10,
    height: 40,
  },
  icon: {
    width: 20,
    height: 20,
    marginRight: 8,
  },
  input: {
    flex: 1,
    fontSize: 16,
    color: '#000',
  },
});

export default React.memo(SearchBar);
