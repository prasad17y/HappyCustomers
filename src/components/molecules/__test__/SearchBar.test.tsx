import React from 'react';
import {render, screen, fireEvent} from '@testing-library/react-native';
import SearchBar from '../SearchBar';

// Tell Jest to use fake timers (for testing debounce)
jest.useFakeTimers();

describe('SearchBar', () => {
  it('should render correctly and match the snapshot', () => {
    const tree = render(
      <SearchBar
        value=""
        onChangeText={jest.fn()}
        placeholder="Search users"
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should call onChangeText with the updated text after the debounce delay', () => {
    const onChangeTextMock = jest.fn();
    const debounceDelay = 500;

    render(
      <SearchBar
        value=""
        onChangeText={onChangeTextMock}
        debounceDelay={debounceDelay}
      />,
    );

    const searchInput = screen.getByPlaceholderText('Search');

    // Simulate user typing "test"
    fireEvent.changeText(searchInput, 'test');

    // At this point, the callback should NOT have been called yet
    expect(onChangeTextMock).not.toHaveBeenCalled();

    // Fast-forward time by the debounce delay
    jest.advanceTimersByTime(debounceDelay);

    // Now, the callback should have been called with the final text
    expect(onChangeTextMock).toHaveBeenCalledWith('test');
    expect(onChangeTextMock).toHaveBeenCalledTimes(1);
  });

  it('should only call onChangeText once for multiple rapid keystrokes', () => {
    const onChangeTextMock = jest.fn();
    const debounceDelay = 500;

    render(
      <SearchBar
        value=""
        onChangeText={onChangeTextMock}
        debounceDelay={debounceDelay}
      />,
    );

    const searchInput = screen.getByPlaceholderText('Search');

    // Simulate rapid typing
    fireEvent.changeText(searchInput, 'a');
    jest.advanceTimersByTime(100); // 100ms
    fireEvent.changeText(searchInput, 'ab');
    jest.advanceTimersByTime(100); // 200ms
    fireEvent.changeText(searchInput, 'abc');
    jest.advanceTimersByTime(100); // 300ms

    // The callback should still not have been called
    expect(onChangeTextMock).not.toHaveBeenCalled();

    // Fast-forward past the debounce timeout
    jest.advanceTimersByTime(debounceDelay);

    // The callback should have been called only once with the final value
    expect(onChangeTextMock).toHaveBeenCalledWith('abc');
    expect(onChangeTextMock).toHaveBeenCalledTimes(1);
  });

  it('should call onDebouncedChange immediately when debounceDelay is 0', () => {
    const onChangeTextMock = jest.fn();
    const debounceDelay = 0;

    render(
      <SearchBar
        value=""
        onChangeText={onChangeTextMock}
        debounceDelay={debounceDelay}
      />,
    );

    const searchInput = screen.getByPlaceholderText('Search');

    // Simulate user typing
    fireEvent.changeText(searchInput, 'instant');

    // The callback should be called immediately without advancing timers
    expect(onChangeTextMock).toHaveBeenCalledWith('instant');
    expect(onChangeTextMock).toHaveBeenCalledTimes(1);
  });
});
