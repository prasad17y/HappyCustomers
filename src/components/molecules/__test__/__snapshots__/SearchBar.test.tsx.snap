// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render correctly and match the snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#f0f2f5",
        "borderRadius": 10,
        "flexDirection": "row",
        "height": 40,
        "paddingHorizontal": 10,
      },
      undefined,
    ]
  }
  testID="-search-bar"
>
  <Image
    source={
      {
        "testUri": "../../../src/assets/search.png",
      }
    }
    style={
      [
        {
          "height": 24,
          "width": 24,
        },
        {
          "height": 20,
          "marginRight": 8,
          "width": 20,
        },
      ]
    }
  />
  <TextInput
    clearButtonMode="while-editing"
    onChangeText={[Function]}
    placeholder="Search users"
    placeholderTextColor="#8e8e93"
    style={
      {
        "color": "#000",
        "flex": 1,
        "fontSize": 16,
      }
    }
    value=""
  />
</View>
`;
